Source/WebCore/rendering/BaselineAlignment.h

```diff
@@ -0,0 +1,126 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "RenderBox.h"
+
+namespace WebCore {
+
+// These classes are used to implement the Baseline Alignment logic, as described in the CSS Box Alignment
+// specification.
+// https://drafts.csswg.org/css-align/#baseline-terms
+//
+// A baseline-sharing group is composed of boxes that participate in baseline alignment together. This is
+// possible only if they:
+//
+//   * Share an alignment context along an axis perpendicular to their baseline alignment axis.
+//   * Have compatible baseline alignment preferences (i.e., the baselines that want to align are on the same
+//     side of the alignment context).
+//
+// Once the BaselineGroup is instantiated, defined by a 'block-direction' (WritingMode) and a 'baseline-preference'
+// (first/last baseline), it's ready to collect the items that will participate in the Baseline Alignment logic.
+//
+class BaselineGroup {
+public:
+    // It stores an item (if not already present) and update the max_ascent associated to this
+    // baseline-sharing group.
+    void update(const RenderBox&, LayoutUnit ascent);
+    LayoutUnit maxAscent() const { return m_maxAscent; }
+    int size() const { return m_items.size(); }
+
+private:
+    friend class BaselineContext;
+    BaselineGroup(WritingMode blockFlow, ItemPosition childPreference);
+
+    // Determines whether a baseline-sharing group is compatible with an item, based on its 'block-flow' and
+    // 'baseline-preference'
+    bool isCompatible(WritingMode, ItemPosition) const;
+
+    // Determines whether the baseline-sharing group's associated block-flow is opposite (LR vs RL) to particular
+    // item's writing-mode.
+    bool isOppositeBlockFlow(WritingMode blockFlow) const;
+
+    // Determines whether the baseline-sharing group's associated block-flow is orthogonal (vertical vs horizontal)
+    // to particular item's writing-mode.
+    bool isOrthogonalBlockFlow(WritingMode blockFlow) const;
+
+    WritingMode m_blockFlow;
+    ItemPosition m_preference;
+    LayoutUnit m_maxAscent;
+    HashSet<const RenderBox*> m_items;
```

Comment:
Can we change this to `HashSet<const RenderBox&>` ?

Reply:
Ah yeah this probably isn't ideal. I wonder if instead we should try to use `WeakPtr` here since the `BaselineAlignment` object doesn't own the renderers?

Reply:
Yeah we should find an alternative to raw pointers here.

Reply:
RenderBox& and RenderBox* really only differ in nullability, they are still both basically bare pointers that can lead to badness. It's better to use a `HashSet<WeakPtr<const RenderBox>>` or maybe a `HashSet<CheckedRef<const RenderBox>>`, which will avoid misuse.